### PR TITLE
[GPT] add sequence_parallel

### PR DESCRIPTION
**Pull request by DrownFish19**
### PR types

New features
### PR changes

Models
### Description

add sequence_parallel

### CODE REVIEW

File: llm/gpt-3/tests/test_sequence_parallel.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        tensor_parallel_rank = mp_group.rank
+
+        if pp_degree > 1:
+            model_class = GPTForCausalLMPipe
+        else:
+            model_class = GPTForCausalLM
+
+        model_name_or_path = "gpt2-medium-en"
+
+        # segment with method: layer:GPTDecoderLayer; result: 0, 13, 27
```
Comment (DesmonDay) [inline llm/gpt-3/tests/test_sequence_parallel.py]: 这块注释看看是否保留

Comment (DrownFish19) [reply llm/gpt-3/tests/test_sequence_parallel.py]: 因为gpt2-medium-en的 ckpt没有LMHead的参数，这个注释特意注明了这一点，推荐保留

File: paddlenlp/transformers/gpt/modeling_pp.py [hunk 2]
```diff
@@ -150,7 +164,7 @@ def __init__(
         config.tensor_parallel_rank = tensor_parallel_rank
 
         self.add_sequential_layer(
-            SharedLayerDesc("gpt", GPTEmbeddingPipe, shared_weight_attr="embedding_weight", config=config),
+            LayerDesc(GPTEmbeddingPipe, config=config),
```
Comment (DesmonDay) [inline paddlenlp/transformers/gpt/modeling_pp.py]: 为啥这块不用Shared了？

File: paddlenlp/transformers/sequence_parallel_utils.py [hunk 3]
```diff
@@ -65,7 +65,9 @@ def all_gather(input):
     output_shape = input.shape
     output_shape[0] = output_shape[0] * parallelism
     output = paddle.empty(shape=output_shape, dtype=input.dtype)
-    group.process_group.all_gather(input, output).wait()
+    # sync between calc_stream and comm_stream will increase execution time
+    # thus use_calc_stream=False
+    dist.stream.all_gather(output, input, group=group, sync_op=True, use_calc_stream=False)
```
Comment (DesmonDay) [inline paddlenlp/transformers/sequence_parallel_utils.py:70]: 这块是因为wait 函数 use_calc_stream默认为True，才改动的么？

File: paddlenlp/transformers/sequence_parallel_utils.py [hunk 4]
```diff
@@ -81,7 +83,9 @@ def reduce_scatter(input):
     )
     output_shape[0] = output_shape[0] // parallelism
     output = paddle.empty(shape=output_shape, dtype=input.dtype)
-    dist.stream.reduce_scatter(output, input, op=dist.ReduceOp.SUM, group=group, sync_op=True)
+    # sync between calc_stream and comm_stream will increase execution time
+    # thus use_calc_stream=False
+    dist.stream.reduce_scatter(output, input, op=dist.ReduceOp.SUM, group=group, sync_op=True, use_calc_stream=False)
```
Comment (DesmonDay) [inline paddlenlp/transformers/sequence_parallel_utils.py]: 不需要修改，默认就是False

File: llm/gpt-3/tests/test_sequence_parallel.py (hunk 1 above)
Comment (DrownFish19) [reply llm/gpt-3/tests/test_sequence_parallel.py]: 之前理解错误，此部分注释已经删除

File: paddlenlp/transformers/gpt/modeling_pp.py (hunk 2 above)
Comment (DrownFish19) [reply paddlenlp/transformers/gpt/modeling_pp.py]: 已经修改为shared

**Review by DesmonDay — APPROVED** [review]
LGTM

File: paddlenlp/transformers/gpt/modeling_pp.py [hunk 5]
```diff
@@ -92,7 +99,19 @@ def forward(self, args):
         input_ids, attention_mask, position_ids = parse_args(args)
         input_ids.stop_gradient = True
         embeddings = super().forward(input_ids=input_ids, position_ids=position_ids)
-        return embeddings
+
+        batch_size, seq_length = input_ids.shape
+        causal_mask = self.bias[:, :, 0:seq_length, :seq_length]
+        if attention_mask is not None:
+            if attention_mask.dtype != paddle.int64:
```
Comment (ZHUI) [inline paddlenlp/transformers/gpt/modeling_pp.py:106]:
```suggestion
        batch_size, seq_length = input_ids.shape
        if attention_mask is not None:
            causal_mask = self.bias[:, :, 0:seq_length, :seq_length]
            if attention_mask.dtype != paddle.int64:
```

File: paddlenlp/transformers/gpt/modeling_pp.py [hunk 6]
```diff
@@ -160,18 +202,8 @@ def __init__(
...
-
         self.add_sequential_layer(
-            SharedLayerDesc(
-                "gpt",
-                GPTEmbeddingPipe,
-                forward_func=_logits_helper,
-                shared_weight_attr="embedding_weight",
-                config=config,
-            ),
+            SharedLayerDesc("gpt", GPTLMHeadPipe, shared_weight_attr="embedding_weight", config=config),
```
Comment (ZHUI) [inline paddlenlp/transformers/gpt/modeling_pp.py:206]: 需要注意正确性，这里的`embedding_weight`是否share成功？

Comment (ZHUI) [reply paddlenlp/transformers/gpt/modeling_pp.py:206]: https://github.com/PaddlePaddle/PaddleFleetX/blob/develop/ppfleetx/models/language_model/gpt/dygraph/hybrid_model.py#L1170-L1180 这里我看 FleetX 没有加在这个位置？

File: paddlenlp/transformers/gpt/modeling_pp.py (hunk 5 above)
Comment (DrownFish19) [reply paddlenlp/transformers/gpt/modeling_pp.py:106]: 1.这里的causal_mask应该在进入if判断之前生成，if和else判断中都会使用causal_mask。
2.参考FleetX代码，也是在进入if判断前生成的
https://github.com/PaddlePaddle/PaddleFleetX/blob/20f33ad21e9d43894b01979dd89d6c32f11e4124/ppfleetx/models/language_model/gpt/dygraph/hybrid_model.py#L865-L878

File: paddlenlp/transformers/gpt/modeling_pp.py (hunk 6 above)
Comment (DrownFish19) [reply paddlenlp/transformers/gpt/modeling_pp.py:206]: 1.数据上确认是正确的，在TP+PP的时候也能share成功的，测试训练几步之后的权重也是一致的。
2.在模型上，head和embedding应该都是直接挂在“gpt”下面的，当前确认是可以加载GPT2的模型文件，FleetX代码挂在‘embed‘，权重文件的命名不同，无法加载GPT2权重。
3.模型上，gpt-modeling_pp.py增加了GPTLMHeadPipe层，将sequence_parallel的处理增加到其中，模型结构更清晰。

File: paddlenlp/transformers/gpt/modeling_pp.py (hunk 5 above)
Comment (ZHUI) [reply paddlenlp/transformers/gpt/modeling_pp.py:106]: 后面那个 causal mask 的分支可以不传入mask，例如使用flash attention的时候不需要。
可以节省一些 pp 间的数据传输
